extensions/wikidata/src/org/openrefine/wikidata/commands/SaveWikibaseSchemaCommand.java

```diff
@@ -55,23 +58,28 @@ public void doPost(HttpServletRequest request, HttpServletResponse response)
             Project project = getProject(request);
 
             String jsonString = request.getParameter("schema");
-            if (jsonString == null) {
+            if (jsonString == null || "null".equals(jsonString)) {
                 respondError(response, "No Wikibase schema provided.");
                 return;
             }
 
             WikibaseSchema schema = ParsingUtilities.mapper.readValue(jsonString, WikibaseSchema.class);
+            
+            ValidationState validation = new ValidationState(project.columnModel);
+            schema.validate(validation);
+            if (!validation.getValidationErrors().isEmpty()) {
+            	Map<String, Object> json = new HashMap<>();
+            	json.put("reason", "invalid-schema");
+            	json.put("message", "Invalid Wikibase schema");
+            	json.put("errors", validation.getValidationErrors());
+				respondJSON(response, json);
+				return;
+            }
 
             AbstractOperation op = new SaveWikibaseSchemaOperation(schema);
             Process process = op.createProcess(project, new Properties());
 
             performProcessAndRespond(request, response, project, process);
-
-        } catch (IOException e) {
-            // We do not use respondException here because this is an expected
-            // exception which happens every time a user tries to save an incomplete
-            // schema - the exception should not be logged.
-            respondError(response, String.format("Wikibase schema could not be parsed: ", e.getMessage()));
```

Comment:
We used to catch all IOExceptions and assume they were due to the schema being invalid, but that was a very sketchy approximation: now schema invalidity errors are handled separately, above.